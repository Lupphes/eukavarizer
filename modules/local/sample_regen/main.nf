process SAMPLE_REHEADER {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/47/474a5ea8dc03366b04df884d89aeacc4f8e6d1ad92266888e7a8e7958d07cde8/data':
        'community.wave.seqera.io/library/bcftools_htslib:0a3fa2654b52006f' }"

    input:
    tuple val(meta), path(vcf)
    val(algorithm)

    output:
    tuple val(meta), path("${prefix}.vcf"), emit: vcf, optional: true
    tuple val(meta), path("${prefix}.vcf.gz"), emit: vcfgz, optional: true
    tuple val(meta), path("${prefix}.vcf.gz.tbi"), emit: tbi, optional: true
    tuple val(meta), path("${prefix}.vcf.gz.csi"), emit: csi, optional: true
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def suffix = task.ext.suffix ?: ''
    prefix = task.ext.prefix ?: "re-${meta.sample}-${meta.platform}-${meta.single_end}-${algorithm}${suffix}"
    def meta_platform = meta.platform ?: "unknown"
    """
    #!/bin/bash
    set -euo pipefail

    # Ensure input is compressed and indexed
    if [[ "${vcf}" == *.vcf.gz ]]; then
        VCF_INPUT="${vcf}"
    else
        VCF_INPUT="input.vcf.gz"
        bgzip --threads ${task.cpus} -c "${vcf}" > "\${VCF_INPUT}"
        tabix --threads ${task.cpus} -p vcf "\${VCF_INPUT}"
    fi

    # Count samples (use || true to handle empty VCFs in strict mode)
    SAMPLE_COUNT=\$(bcftools query -l "\${VCF_INPUT}" 2>/dev/null | wc -l || true)

    if [[ "\${SAMPLE_COUNT}" -eq 1 ]]; then
        # Rename sample using bcftools reheader
        echo "${meta.sample}" > new_sample.txt
        bcftools reheader -s new_sample.txt "\${VCF_INPUT}" -o temp_reheadered.vcf.gz

        # Stream processing - add INFO headers and values
        {
            # Output all header lines except #CHROM line
            bcftools view --no-version -h temp_reheadered.vcf.gz | grep -v "^#CHROM"

            # Add custom INFO headers
            cat <<'INFO_HEADERS'
##INFO=<ID=EUK_CALLER,Number=1,Type=String,Description="SV calling algorithm used">
##INFO=<ID=EUK_PLATFORM,Number=1,Type=String,Description="Sequencing platform">
##INFO=<ID=EUK_SE,Number=1,Type=String,Description="Single-end or paired-end">
INFO_HEADERS

            # Output #CHROM line
            bcftools view --no-version -h temp_reheadered.vcf.gz | grep "^#CHROM"

            # Add INFO values to each variant
            bcftools view --no-version -H temp_reheadered.vcf.gz | awk -v algo="${algorithm}" -v plat="${meta_platform}" -v se="${meta.single_end}" \\
                'BEGIN {FS=OFS="\\t"} {
                    info_add = "EUK_CALLER=" algo ";EUK_PLATFORM=" plat ";EUK_SE=" se
                    if (\$8 == "." || \$8 == "") {
                        \$8 = info_add
                    } else {
                        \$8 = \$8 ";" info_add
                    }
                    print
                }'
        } | bgzip --threads ${task.cpus} > ${prefix}.vcf.gz

        # Index the output
        tabix --threads ${task.cpus} -p vcf ${prefix}.vcf.gz
        bcftools index --threads ${task.cpus} --csi ${prefix}.vcf.gz

        # Create uncompressed version
        bcftools view --no-version -Ov -o ${prefix}.vcf ${prefix}.vcf.gz

        # Cleanup
        rm -f temp_reheadered.vcf.gz new_sample.txt

    elif [[ "\${SAMPLE_COUNT}" -eq 0 ]]; then
        echo "WARNING: No samples found in VCF (sites-only). Skipping reheadering." >&2

    else
        echo "ERROR: VCF contains \${SAMPLE_COUNT} samples (expected 1). Cannot proceed." >&2
        exit 1
    fi

    # Capture versions
    cat <<EOF > versions.yml
"${task.process}":
    bcftools: \$(bcftools --version 2>&1 | head -n 1 | sed 's/^.*bcftools //')
    tabix: \$(tabix --version 2>&1 | head -n 1 | sed 's/^.*tabix (htslib) //' | sed 's/ .*//')
EOF
    """

    stub:
    def suffix = task.ext.suffix ?: ''
    prefix = task.ext.prefix ?: "re-${meta.sample}-${meta.platform}-${meta.single_end}-${algorithm}${suffix}"
    """
    touch ${prefix}.vcf
    echo -n | gzip > ${prefix}.vcf.gz
    touch ${prefix}.vcf.gz.tbi
    touch ${prefix}.vcf.gz.csi
    touch versions.yml
    """
}
