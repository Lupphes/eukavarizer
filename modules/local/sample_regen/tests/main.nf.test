nextflow_process {

    name "Test Process SAMPLE_REHEADER"
    script "../main.nf"
    process "SAMPLE_REHEADER"
    tag "sample_regen"
    tag "modules"
    tag "modules_local"

    test("Reheader single-sample VCF - success") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = channel.of([
                    [ id: 'test', sample: 'test_sample', platform: 'illumina', single_end: false ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf')
                ])
                input[1] = "manta"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.vcf || process.out.vcfgz },
                { assert process.out.versions }
            )
        }
    }

    test("Reheader compressed VCF with index") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = channel.of([
                    [ id: 'test2', sample: 'sample2', platform: 'ont', single_end: true ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf.gz')
                ])
                input[1] = "sniffles"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.vcfgz },
                { assert process.out.tbi },
                { assert process.out.csi },
                { assert process.out.versions }
            )
        }
    }

    test("Stub test") {

        options '-stub'

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = channel.of([
                    [ id: 'stub', sample: 'stub_sample', platform: 'pacbio', single_end: false ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/vcf/test.rnaseq.vcf')
                ])
                input[1] = "delly"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert process.out.versions }
            )
        }
    }
}
