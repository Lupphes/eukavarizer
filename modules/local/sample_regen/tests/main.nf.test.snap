{
    "Stub test": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "csi": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "tbi": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcfgz": [
                    [
                        {
                            "id": "stub",
                            "sample": "stub_sample",
                            "platform": "pacbio",
                            "single_end": false
                        },
                        "re-stub_sample-pacbio-false-delly.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-18T23:10:52.734336894"
    },
    "Reheader compressed VCF with index": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf:md5,56ef861130b57aebdf42fc1afd7507a2"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf.gz:md5,56ef861130b57aebdf42fc1afd7507a2"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf.gz.tbi:md5,d71d6955f3d7b8b874de8b80690030eb"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf.gz.csi:md5,5b108b4b1a6ab405cae4a1f960f5ce34"
                    ]
                ],
                "4": [
                    "versions.yml:md5,a1ede0be0456a875fcae37e667abc2c6"
                ],
                "csi": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf.gz.csi:md5,5b108b4b1a6ab405cae4a1f960f5ce34"
                    ]
                ],
                "tbi": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf.gz.tbi:md5,d71d6955f3d7b8b874de8b80690030eb"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf:md5,56ef861130b57aebdf42fc1afd7507a2"
                    ]
                ],
                "vcfgz": [
                    [
                        {
                            "id": "test2",
                            "sample": "sample2",
                            "platform": "ont",
                            "single_end": true
                        },
                        "re-sample2-ont-true-sniffles.vcf.gz:md5,56ef861130b57aebdf42fc1afd7507a2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a1ede0be0456a875fcae37e667abc2c6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-18T23:03:57.837531726"
    },
    "Reheader single-sample VCF - success": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf:md5,395e8aae23869e2118229c8e4f869607"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf.gz:md5,395e8aae23869e2118229c8e4f869607"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf.gz.tbi:md5,af64b769e31b2c6d20dd1a9ca5b422dd"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf.gz.csi:md5,1ccfe8edfdae619a9f98f91a27718608"
                    ]
                ],
                "4": [
                    "versions.yml:md5,a1ede0be0456a875fcae37e667abc2c6"
                ],
                "csi": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf.gz.csi:md5,1ccfe8edfdae619a9f98f91a27718608"
                    ]
                ],
                "tbi": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf.gz.tbi:md5,af64b769e31b2c6d20dd1a9ca5b422dd"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf:md5,395e8aae23869e2118229c8e4f869607"
                    ]
                ],
                "vcfgz": [
                    [
                        {
                            "id": "test",
                            "sample": "test_sample",
                            "platform": "illumina",
                            "single_end": false
                        },
                        "re-test_sample-illumina-false-manta.vcf.gz:md5,395e8aae23869e2118229c8e4f869607"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a1ede0be0456a875fcae37e667abc2c6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-18T23:03:48.608838479"
    }
}